Optimizes serialMaster

JasonWong08 opened the pull request:
add module SerialCommunication.py
add demos/cuteNybble.py
optimizes ardSerial.py
modifies example.py (add  more comments)
deleted ardSerial.pyc, step.py, test.py, test_File.py

borntoleave commented:
Thanks for your suggestion! I love the board manager solution and will test it after I ship out the current batch of orders!
BTW, is it possible to add more information in the board information? For example, there will be NyBoard_V0_1 and NyBoard_V0_2, some of the pin definitions will be different. Is it possible to change them automatically when selecting boards?

asmagill commented:
I'm pretty new to crafting this kind of thing, so let me qualify what follows with "but I'd need to create some test cases to be certain"...

I do know you can include multiple boards in the files and that you can pass in specific compiler flags based on the board chosen with lines like `XXX.build.extra_flags=-DNYBOARD01` (allowing things like `#ifdef NUBOARD01` in code)

I'm not sure, but I *think* the `XXX.build.variant` line allows you to specify an include file that automatically gets included when building the sketch as well, so you could specify more complex code differences that way.

It was helpful to me when crafting this pull to review some of the other 3rd party definitions and see how they put things together -- you can find unofficial list of these at https://github.com/arduino/Arduino/wiki/Unofficial-list-of-3rd-party-boards-support-urls.

borntoleave commented:
Hi, sorry for delayed response. I just validated your changes and it works seamlessly on Linux and Mac. 
Next I'll need to figure out how to add "#define" conditions to change parameters for different boards. 

Thank you for your help!

JasonWong08 committed:
Optimizes serialMaster

add module SerialCommunication.py
add demos/cuteNybble.py
optimizes ardSerial.py
modifies example.py (add  more comments)
deleted ardSerial.pyc, step.py, test.py, test_File.py
